WikiWatershed/model-my-watershed
# Compare View: TR-55 Runoff Charts

## Overview

This PR adds the TR-55 Compare View charts and sets up the table & chart values to adjust accordingly. gif demo's below.

Connects #2072 

### Demo

![compare-charts](https://user-images.githubusercontent.com/4165523/29145008-25e08422-7d29-11e7-9000-d3097af7a95a.gif)

### Notes

Currently there are a few quirks & limitations with this implementation, which we can either handle here or spin off elsewhere. 

1. I did not implement the popovers, which seem to be complicated enough to handle in another card.  NVD3 has tooltips by default, but since the desired behavior's that the hovering over one value on the bar chart opens corresponding popovers for all three, which isn't a built-in NVD3 feature. I think we may end up wanting to use Bootstrap popovers for this.

2. The bars themselves match the width of the scenario title placeholders when there are few enough of them to display in the scenario header without overflow. However, when there are more, the chart div seems not to expand in size beyond what's visible so the bars get squished.

3. There's currently some flicker as new charts are shown when retrieving any of the results, which means it renders once per scenario if the precipitation slider value changes. Not sure what is the best way to handle this -- I think we may want to set some value to check whether any of the scenarios is currently polling and show a loading state if not?

4. I think the maximum y axis value for "Combined Hydrology" should be the value the precipitation slider has been set to, but it seemed like the value these scenarios had was slightly off the number displayed for the slider. Not sure what is the best way to handle this, but we'll need to pass it in to the chart to set the `yDomain`.

Happy to try to address these here or to spin them off elsewhere; figured I'd put up the PR for review rather than spinning wheels on it.

## Testing
- get this branch, then ~`update` and `server`~ `bundle.sh` & hard refresh in the browser
- create a new TR-55 project then create a few scenarios including some which would cause varied values for runoff/infiltration (think: veg basin gswi conservation practice or high density land cover)
- open the compare view and verify that you see the new charts and that the values seem accurate to the underlying data
- move the precipitation slider and verify that when the polling stops, the values in the chart have changed
- click to see the table view, then move the precipitation slider, and verify that those values change too
